test/prototype/test_parq.py

```diff
@@ -53,52 +78,184 @@ def forward(self, x):
+    def test_int4_weight_only(self, group_size: int = 32):
+        model = M(m=512, n=512).to(torch.bfloat16).to(_DEVICE)
+        model.reset_parameters()
+
+        m_ref = copy.deepcopy(model)
+        quantize_(m_ref, Int4WeightOnlyConfig(group_size))
+
+        b = 4
+        quantizer = self.int4_torchao_quantizer()
+        self.compare_quantized_models(model, m_ref, quantizer, b, group_size)
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_6, "Test only enabled for 2.6+")
+    @common_utils.parametrize("b", [2, 3, 4, 8])
+    @common_utils.parametrize("group_size", [32, 512])
+    def test_intx_weight_only(self, b: int = 2, group_size: int = 32):
+        model = M(m=512, n=512).to(_DEVICE)
+        model.reset_parameters()
+
+        m_ref = copy.deepcopy(model)
+        quantize_(
+            m_ref,
+            IntxWeightOnlyConfig(
+                weight_dtype=_BIT_WIDTH_TO_DTYPE[b], granularity=PerGroup(group_size)
+            ),
+        )
+
+        quantizer = UnifTorchaoQuantizer(
+            symmetric=True,
+            preserve_zero=True,
+            zero_point_domain=ZeroPointDomain.INT,
+        )
+        self.compare_quantized_models(model, m_ref, quantizer, b, group_size)
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_4, "Test only enabled for 2.4+")
+    @unittest.skipIf(_DEVICE == "cpu", "Need GPU available")
+    def test_int4_weight_only_e2e(self, group_size: int = 32):
+        model = M(m=512, n=512).to(torch.bfloat16).to(_DEVICE)
+        model.reset_parameters()
+
+        m_ref = copy.deepcopy(model)
+        config = Int4WeightOnlyConfig(group_size)
+        quantize_(m_ref, config)
+
+        b = 4
+        params_quant, params_no_quant = split_param_groups(model)
+        param_groups = [
+            {"params": params_quant, "quant_bits": b, "quant_block_size": group_size},
+            {"params": params_no_quant},
+        ]
+        base_optimizer = torch.optim.AdamW(param_groups)
+
+        quantizer = self.int4_torchao_quantizer(config=config)
+        optimizer = QuantOptimizer(
+            base_optimizer, quantizer, ProxHardQuant(), quant_per_channel=True
+        )
+
+        # do not update model weights, just quantize
+        optimizer.zero_grad()
+        optimizer.step()
+
+        orig_model = copy.deepcopy(model)  # save copy of PARQ quantized model
+
+        # equivalent to torchao's convert step
+        with torch.no_grad():
+            optimizer.restore_latent_params()
+        quantize_(model, quantizer.config)
+
+        for n, module in model.named_modules():
+            if not _is_linear(module):
+                continue
+
+            p_orig = getattr(orig_model, n).weight  # PARQ weight
+            p = module.weight.dequantize()  # PARQ weight after quantize_
+            p_ref = getattr(m_ref, n).weight.dequantize()  # natively quantize_
+
+            self.assertTrue(p_orig.equal(p_ref))
+            self.assertTrue(p.equal(p_ref))
```

Comment:
@metascroy Lines 241–243 shows how to convert a PARQ quantized model using `quantize_`. Since PARQ stores full-precision "latent" copies of each quantized parameter, I just load those values into the model before calling `quantize_`. Line 254 shows that we match the natively quantized model `m_ref`.